map/index.d.ts

```diff
@@ -89,7 +89,7 @@ export interface MapStore<Value extends object = any>
    */
   setKey<Key extends AllKeys<Value>>(
     key: Key,
-    value: Get<Value, Key> | Value[Key]
+    value: Get<Value, Key> | undefined | Value[Key]
```

Comment:
This change will allow removing key which it is prohibited to remove:

```ts
let $store = map<{ required: string, optional?: string }>({ required: 'a' })
$store.setKey('optional', undefined) // OK
$store.setKey('required', undefined) // Expected to throw
```

You should not use just `Record` if you allow removing any key.

Reply:
Yeah, you are right, it breaks types for literal objects...

> You should not use just Record if you allow removing any key.

What do you mean, can you elaborate? I have such code:
```
type Id = string;
type HeightMap = Record<Id, number>

const $heightMap = map<HeightMap>();
```

And it doesn't allow setting `undefined` to remove key from map.
```
// Argument of type `undefined` is not assignable to parameter of type `number`
$heightMap.setKey('id', undefined)
```

Should I use `map` here at all?

Reply:
I think will solve the problem `{ [key: string]?: number }`

Reply:
This will break types in other places 🙂. For example, you have helper for getting total height from map:
```
function calcTotalHeight(heightMap: HeightMap) {
   return Object.values(heightMap).reduce((acc, height) => acc + height, 0);
}
```
With your suggestion height might be undefined, but in reality it can't. [Proof](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAkgJlAvFAzsATgSwHYHMDcAUKJFABISa4AWwAsgIZhJQBKEAxgPbpwA88ADRRsAVwC2AIwjoAfERLQKVWozAB5MMExdsDADYsACg3TaDfZTXpNZhQgDNR2Dtt1QOBjgBUuwA1a0ABTUlNZqAFzkYapMAJRQAN6EUKnoEMCi6NhQ6pIAVpzAAHQAbgaiECghMTZgccXpcKIcEEFBDBwcwqEqwAmIslCdHFAA1FC91sIADHFEAL72Ti5uOZ76Pn4BtZprBjV9kdFHTHs6evoJyalQ6ZnZuQVFZRVVh+HxjRDNre0jPVqAyGI3Gk1qs3mhAWQA)

It looks like the root of problem in the ambiguity of the `setKey` method:
1. It considers `undefined` as special value and doesn't set key, but removes it instead (which is contradicts to method name, btw)
2. It doesn't allow `undefined` as values at all.

Have you thought about separate method for removing key, like `removeKey/deleteKey`? It looks like it would solve all these issues

Reply:
What type you see for `removeKey`?

Reply:
After some investigations I understood that it will have the same issue, because `Record` type is just syntax sugar for object literal, and it seems it isn't possible to differentiate them. 
Ideologically,  I think the current `.map` implementation is more likely `.object` or `.record`, and `.map` must work with `Record` type instead, which is basically map between key and value.
I don't think you want to change API in this way, so I can just drop that commit and leave the other one. Wdyt?